### ISSUE TITLE
🐛 Bug Report: Cannot set headers after they are sent to the client error using enableExperimentalRedirectFlow

### ISSUE THREAD

**Issue by awanlin** [backstage/backstage#18930]
### 📜 Description

After we enabled the `enableExperimentalRedirectFlow` we've started to see the error below in our logs. This error happens right after someone logs in. So far from what we can tell this hasn't impacted anything but it's worrisome. We see this happen with both a custom auth provider as well as the Azure auth provider, we are fairly confident this will happen with all of them.

```text
[1] 2023-08-03T15:34:23.132Z backstage error Request failed with status 500 Cannot set headers after they are sent to the client type=errorHandler code=ERR_HTTP_HEADERS_SENT stack=Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
[1]     at new NodeError (node:internal/errors:387:5)
[1]     at ServerResponse.setHeader (node:_http_outgoing:603:11)
[1]     at postMessageResponse (/Users/andre.wanlin/repos/github/x/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:269:7)
[1]     at OAuthAdapter.frameHandler (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:439:14)
[1]     at async OAuthEnvironmentHandler.frameHandler (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:211:5)
[1] Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
[1]     at new NodeError (node:internal/errors:387:5)
[1]     at ServerResponse.setHeader (node:_http_outgoing:603:11)
[1]     at postMessageResponse (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:269:7)
[1]     at OAuthAdapter.frameHandler (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:439:14)
[1]     at async OAuthEnvironmentHandler.frameHandler (/Users/andre.wanlin/repos/githubbackstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:211:5)
```

### 👍 Expected behavior

There should be no errors after logging in using the `enableExperimentalRedirectFlow`

### 👎 Actual Behavior with Screenshots

After logging in using the `enableExperimentalRedirectFlow` we get this error:

```text
[1] 2023-08-03T15:34:23.132Z backstage error Request failed with status 500 Cannot set headers after they are sent to the client type=errorHandler code=ERR_HTTP_HEADERS_SENT stack=Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
[1]     at new NodeError (node:internal/errors:387:5)
[1]     at ServerResponse.setHeader (node:_http_outgoing:603:11)
[1]     at postMessageResponse (/Users/andre.wanlin/repos/github/x/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:269:7)
[1]     at OAuthAdapter.frameHandler (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:439:14)
[1]     at async OAuthEnvironmentHandler.frameHandler (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:211:5)
[1] Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
[1]     at new NodeError (node:internal/errors:387:5)
[1]     at ServerResponse.setHeader (node:_http_outgoing:603:11)
[1]     at postMessageResponse (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:269:7)
[1]     at OAuthAdapter.frameHandler (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:439:14)
[1]     at async OAuthEnvironmentHandler.frameHandler (/Users/andre.wanlin/repos/githubbackstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:211:5)
```

### 👟 Reproduction steps

1. Create a Backstage instance
2. Setup the Azure auth provider
3. In your `app-config.yaml` add the following at the root: `enableExperimentalRedirectFlow: true`
4. Start up Backstage
5. Log in

Notice in the logs you will see an error like this:

```text
[1] 2023-08-03T15:34:23.132Z backstage error Request failed with status 500 Cannot set headers after they are sent to the client type=errorHandler code=ERR_HTTP_HEADERS_SENT stack=Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
[1]     at new NodeError (node:internal/errors:387:5)
[1]     at ServerResponse.setHeader (node:_http_outgoing:603:11)
[1]     at postMessageResponse (/Users/andre.wanlin/repos/github/x/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:269:7)
[1]     at OAuthAdapter.frameHandler (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:439:14)
[1]     at async OAuthEnvironmentHandler.frameHandler (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:211:5)
[1] Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
[1]     at new NodeError (node:internal/errors:387:5)
[1]     at ServerResponse.setHeader (node:_http_outgoing:603:11)
[1]     at postMessageResponse (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:269:7)
[1]     at OAuthAdapter.frameHandler (/Users/andre.wanlin/repos/github/backstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:439:14)
[1]     at async OAuthEnvironmentHandler.frameHandler (/Users/andre.wanlin/repos/githubbackstage/node_modules/@backstage/plugin-auth-backend/dist/index.cjs.js:211:5)
```

### 📃 Provide the context for the Bug.

We have a requirement to not have a pop-up as all our other applications do not. Because of this the `enableExperimentalRedirectFlow` flag is a great solution and seems to work exactly as we would like other than this error 👍 

### 🖥️ Your Environment

```text
OS:   Darwin 22.6.0 - darwin/arm64
node: v18.13.0
yarn: 3.6.1
cli:  0.22.10-next.1 (installed)
backstage:  1.17.0-next.1

Dependencies:
  @backstage/app-defaults                                            1.4.2-next.0
  @backstage/backend-app-api                                         0.5.0-next.1
  @backstage/backend-common                                          0.19.2-next.1
  @backstage/backend-dev-utils                                       0.1.1
  @backstage/backend-openapi-utils                                   0.0.3-next.1
  @backstage/backend-plugin-api                                      0.6.0-next.1
  @backstage/backend-tasks                                           0.5.5-next.1
  @backstage/backend-test-utils                                      0.2.0-next.1
  @backstage/catalog-client                                          1.4.3
  @backstage/catalog-model                                           1.4.1
  @backstage/cli-common                                              0.1.12
  @backstage/cli-node                                                0.1.3-next.0
  @backstage/cli                                                     0.22.10-next.1
  @backstage/config-loader                                           1.4.0-next.1
  @backstage/config                                                  1.0.8
  @backstage/core-app-api                                            1.9.1-next.0
  @backstage/core-components                                         0.13.3, 0.13.4-next.0
  @backstage/core-plugin-api                                         1.5.3
  @backstage/dev-utils                                               1.0.18-next.1
  @backstage/errors                                                  1.2.1
  @backstage/eslint-plugin                                           0.1.3
  @backstage/integration-aws-node                                    0.1.5
  @backstage/integration-react                                       1.1.16-next.1
  @backstage/integration                                             1.5.1
  @backstage/plugin-adr-backend                                      0.3.6-next.1
  @backstage/plugin-adr-common                                       0.2.11
  @backstage/plugin-adr                                              0.6.4-next.1
  @backstage/plugin-api-docs                                         0.9.7-next.1
  @backstage/plugin-app-backend                                      0.3.48-next.1
  @backstage/plugin-auth-backend                                     0.18.6-next.1
  @backstage/plugin-auth-node                                        0.2.17-next.1
  @backstage/plugin-azure-devops-backend                             0.3.27-next.1
  @backstage/plugin-azure-devops-common                              0.3.0
  @backstage/plugin-azure-devops                                     0.3.3-next.0
  @backstage/plugin-badges-backend                                   0.2.3-next.1
  @backstage/plugin-badges                                           0.2.45-next.0
  @backstage/plugin-catalog-backend-module-bitbucket-server          0.1.13-next.1
  @backstage/plugin-catalog-backend-module-github                    0.3.3-next.1
  @backstage/plugin-catalog-backend                                  1.12.0-next.1
  @backstage/plugin-catalog-common                                   1.0.15
  @backstage/plugin-catalog-graph                                    0.2.33-next.1
  @backstage/plugin-catalog-import                                   0.9.11-next.1
  @backstage/plugin-catalog-node                                     1.4.1-next.1
  @backstage/plugin-catalog-react                                    1.8.0, 1.8.1-next.0
  @backstage/plugin-catalog                                          1.12.1-next.1
  @backstage/plugin-entity-feedback-backend                          0.1.6-next.1
  @backstage/plugin-entity-feedback-common                           0.1.2-next.0
  @backstage/plugin-entity-feedback                                  0.2.4-next.1
  @backstage/plugin-events-backend-module-aws-sqs                    0.2.3-next.1
  @backstage/plugin-events-backend-module-azure                      0.1.10-next.1
  @backstage/plugin-events-backend-module-bitbucket-cloud            0.1.10-next.1
  @backstage/plugin-events-backend-module-github                     0.1.10-next.1
  @backstage/plugin-events-backend                                   0.2.9-next.1
  @backstage/plugin-events-node                                      0.2.9-next.1
  @backstage/plugin-explore-common                                   0.0.1
  @backstage/plugin-explore-react                                    0.0.30
  @backstage/plugin-explore                                          0.4.7-next.0
  @backstage/plugin-github-actions                                   0.6.2-next.1
  @backstage/plugin-github-pull-requests-board                       0.1.15-next.0
  @backstage/plugin-home-react                                       0.1.2-next.0
  @backstage/plugin-home                                             0.5.5-next.0
  @backstage/plugin-kubernetes-backend                               0.11.3-next.1
  @backstage/plugin-kubernetes-common                                0.6.5
  @backstage/plugin-kubernetes                                       0.9.4-next.0
  @backstage/plugin-org                                              0.6.11-next.1
  @backstage/plugin-permission-backend                               0.5.23-next.1
  @backstage/plugin-permission-common                                0.7.7
  @backstage/plugin-permission-node                                  0.7.11-next.1
  @backstage/plugin-permission-react                                 0.4.14
  @backstage/plugin-playlist-backend                                 0.3.4-next.1
  @backstage/plugin-playlist-common                                  0.1.9-next.0
  @backstage/plugin-playlist                                         0.1.13-next.1
  @backstage/plugin-proxy-backend                                    0.2.42-next.1
  @backstage/plugin-scaffolder-backend-module-confluence-to-markdown 0.2.1-next.1
  @backstage/plugin-scaffolder-backend                               1.15.2-next.1
  @backstage/plugin-scaffolder-common                                1.3.2
  @backstage/plugin-scaffolder-node                                  0.1.6-next.1
  @backstage/plugin-scaffolder-react                                 1.5.2-next.0
  @backstage/plugin-scaffolder                                       1.14.2-next.1
  @backstage/plugin-search-backend-module-catalog                    0.1.4-next.1
  @backstage/plugin-search-backend-module-elasticsearch              1.3.3-next.1
  @backstage/plugin-search-backend-module-pg                         0.5.9-next.1
  @backstage/plugin-search-backend-module-techdocs                   0.1.4-next.1
  @backstage/plugin-search-backend-node                              1.2.4-next.1
  @backstage/plugin-search-backend                                   1.4.0-next.1
  @backstage/plugin-search-common                                    1.2.5
  @backstage/plugin-search-react                                     1.6.4-next.0
  @backstage/plugin-search                                           1.3.4-next.0
  @backstage/plugin-shortcuts                                        0.3.13-next.1
  @backstage/plugin-sonarqube-backend                                0.2.2-next.1
  @backstage/plugin-sonarqube-react                                  0.1.7
  @backstage/plugin-sonarqube                                        0.7.2-next.1
  @backstage/plugin-tech-insights-backend-module-jsonfc              0.1.32-next.1
  @backstage/plugin-tech-insights-backend                            0.5.14-next.1
  @backstage/plugin-tech-insights-common                             0.2.11
  @backstage/plugin-tech-insights-node                               0.4.6-next.1
  @backstage/plugin-tech-insights                                    0.3.13-next.1
  @backstage/plugin-tech-radar                                       0.6.7-next.0
  @backstage/plugin-techdocs-backend                                 1.6.5-next.1
  @backstage/plugin-techdocs-module-addons-contrib                   1.0.16-next.1
  @backstage/plugin-techdocs-node                                    1.7.4-next.1
  @backstage/plugin-techdocs-react                                   1.1.9-next.0
  @backstage/plugin-techdocs                                         1.6.6-next.1
  @backstage/plugin-user-settings                                    0.7.6-next.0
  @backstage/release-manifests                                       0.0.9
  @backstage/test-utils                                              1.4.2-next.0
  @backstage/theme                                                   0.4.1
  @backstage/types                                                   1.1.0
  @backstage/version-bridge                                          1.0.4
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by awanlin** [comment]
Related information:

- PR that added this feature: https://github.com/backstage/backstage/pull/15841
- Issue that requested this feature: https://github.com/backstage/backstage/issues/9582

**Comment by awanlin** [comment]
@headphonejames @Rugvip just wondering if you have any thoughts on this?

I think the code that's triggering the error is here, but I'm not sure about why this is happening:

https://github.com/backstage/backstage/blob/e8bebb6f5f7e20539511720778afc478510a5cc9/plugins/auth-backend/src/lib/flow/authFlowHelpers.ts#L66-L69

**Comment by HChenCBRE** [comment]
https://github.com/backstage/backstage/blob/e8bebb6f5f7e20539511720778afc478510a5cc9/plugins/auth-backend/src/lib/oauth/OAuthAdapter.ts#L183-L190
The postMessageResponse is called after the redirect might caused this

**Comment by freben** [comment]
Awesome, closed by #19232, thanks @awanlin!

**Comment by awanlin** [comment]
Thanks for the pointer @HChenCBRE! 🚀
